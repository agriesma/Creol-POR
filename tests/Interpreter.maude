***
*** Interpreter.maude -- Test cases for the interpreter.
***
*** Copyright (c) 2007
***
*** This program is free software; you can redistribute it and/or
*** modify it under the terms of the GNU General Public License as
*** published by the Free Software Foundation; either version 2 of the
*** License, or (at your option) any later version.
***
*** This program is distributed in the hope that it will be useful, but
*** WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
*** General Public License for more details.
***
*** You should have received a copy of the GNU General Public License
*** along with this program; if not, write to the Free Software
*** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
*** 02111-1307, USA.
***

set show stats off .
set show timing off .
set show advise off .
load interpreter .

mod TEST-TAILCALL is
  pr INTERPRETER .
  op init : -> Configuration .
  eq init =

    < 'Tailcall : Cl | Inh: noInh, Par: 'x, Att: 'x |-> null, Mtds:
	< 'init : Mtdname | Param: noAid, Latt: noSubst, Code: return(emp) > *
	< 'run : Mtdname | Param: noAid, Latt: noSubst, Code:
          if 'less[[ int(0) # 'x ]] th 'x ::= ('minus[[ 'x # int(1) ]]) ;
	    tailcall 'run(emp) el return(emp) fi >,
      Ocnt: 0 >

  main('Tailcall, int(5)) .

endm

red in TEST-TAILCALL : init .

search [1] in TEST-TAILCALL : init =>+
  < ob('Tailcall0) : 'Tailcall | Att: A:Subst, Pr:
    L:Subst, accept(tag(label(ob('Tailcall0), 2))) ; SL:StmList, PrQ: W:MProc, Lcnt: N:Nat >
  conf:Configuration .

rew in TEST-TAILCALL : init .

quit
