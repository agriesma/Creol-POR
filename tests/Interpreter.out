==========================================
reduce in TEST-TAILCALL : init .
result Configuration: 
< "Tailcall" : Class | Inh: noInh, Param: "x", Att: "x" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: noSubst, Code: return(emp) >, < "run"
    : Method | Param: noVid, Att: noSubst, Code: if "<"(int(0) :: "x") th
    assign("x" ; "-"("x" :: int(1))) ; tailcall("run" ; emp) el return(emp) fi
    >,
  Ocnt: 0 >
 
< ob("main") : "" |
  Att: noSubst,
  Pr: {"var" |-> null | new("var" ; "Tailcall" ; int(5))},
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 0 >

==========================================
search in TEST-TAILCALL : init =>+ conf:Configuration 
< ob("Tailcall0") : "Tailcall" |
  Att: A:Subst,
  Pr: {L:Subst | $accept(tag(label(ob("Tailcall0"), 1))) ; SL:StmtList},
  PrQ: W:MProc,
  Dealloc: LS:Labels,
  Ev: MM:MMsg,
  Lcnt: N:Nat >
 .

Solution 1 (state 22)
conf:Configuration --> 
< "Tailcall" : Class | Inh: noInh, Param: "x", Att: "x" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: noSubst, Code: return(emp) >, < "run"
    : Method | Param: noVid, Att: noSubst, Code: if "<"(int(0) :: "x") th
    assign("x" ; "-"("x" :: int(1))) ; tailcall("run" ; emp) el return(emp) fi
    >,
  Ocnt: 1 >
 
< ob("main") : "" |
  Att: noSubst,
  Pr: {"var" |-> null | assign("var" ; ob("Tailcall0"))},
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 0 >

A:Subst --> "this" |-> ob("Tailcall0"), "x" |-> int(4)
L:Subst --> noSubst
SL:StmtList --> noStmt
W:MProc --> {".class" |-> str("Tailcall"), ".label" |-> tag(label(ob(
    "Tailcall0"), 1)), ".method" |-> str("run"), "caller" |-> ob("Tailcall0") |
    if "<"(int(0) :: "x") th assign("x" ; "-"("x" :: int(1))) ; tailcall("run"
    ; emp) el return(emp) fi}
LS:Labels --> label(ob("Tailcall0"), 1)
MM:MMsg --> noMsg
N:Nat --> 2
==========================================
search in TEST-TAILCALL : init =>+ conf:Configuration 
< ob("Tailcall0") : "Tailcall" |
  Att: A:Subst,
  Pr: {L:Subst | $accept(tag(label(ob("Tailcall0"), 2))) ; SL:StmtList},
  PrQ: W:MProc,
  Dealloc: LS:Labels,
  Ev: MM:MMsg,
  Lcnt: N:Nat >
 .

No solution.
==========================================
rewrite in TEST-TAILCALL : init .
result Configuration: 
< "Tailcall" : Class | Inh: noInh, Param: "x", Att: "x" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: noSubst, Code: return(emp) >, < "run"
    : Method | Param: noVid, Att: noSubst, Code: if "<"(int(0) :: "x") th
    assign("x" ; "-"("x" :: int(1))) ; tailcall("run" ; emp) el return(emp) fi
    >,
  Ocnt: 1 >
 
< ob("Tailcall0") : "Tailcall" |
  Att: "this" |-> ob("Tailcall0"), "x" |-> int(0),
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 2 >
 
< ob("main") : "" |
  Att: noSubst,
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 0 >

Bye.
